### PR TITLE
feat(mgnp): initial MGNP protocol implementation

### PR DESCRIPTION
**Pull request by hawkw**
This branch adds a rough, but basically working, implementation of an
initial MGNP protocol library, including a connection-management state
machine and routing. I've written a single integration test with two
peers running in separate Tokio tasks and communicating over a Tokio
MPSC of frames, where one peer connects to a "hello world" service on
the remote and they exchange "hello world" messages with each other, and
that seems to basically work.

Some things that aren't done yet include:
- [_] More tests, obviously, especially around the various edge cases of
  creating/closing connections.
- [_] Currently, when a local connect is initiated, the client just
  immediately gets a BiDi, regardless of whether the connect succeeds or
  fails. We need to add a oneshot channel type thing so that the client
  can wait for the connection to be handshaked and receive an error if
  the handshake failed, but I haven't done that yet.
- [_] The registry service identity design is a kind of rough sketch
  that needs more love.
- [_] The interface exposed by outbound frames to the `Wire` is quite
  awkward, and needs to be reworked to make writing out serialized
  messages easier. This is a bit tough with the current `tricky-pipe`
  and `postcard` interfaces.

I'd prefer to save all of this work for follow-up branches and get the
current working impl merged, so that we can move forward with what we
have currently.

### CODE REVIEW

File: source/mgnp-pitch/src/lib.rs [hunk 1]
```diff
@@ -29,6 +47,48 @@ enum IdentityKind {
...
+    pub async fn handle_connection(&mut self) -> Result<(), ()> {
+        let Self { wire, conn_table } = self;
+        loop {
+            let frame = futures::select_biased! {
+                // inbound frame from the wire.
+                frame = wire.recv().fuse() => frame?,
+                // either a connection needs to send data, or a connection has
+                // closed locally.
+                frame = conn_table.next_outbound().fuse() => {
+                    let frame: Fr = todo!("eliza: serialize outbound msg to frame; may need interface changes...");
```
Comment (jamesmunns) [inline source/mgnp-pitch/src/lib.rs:67]: Not sure I totally understand, but a bidi tricky pipe should handle this.

The conn table will hold `Bidi<Frame, Frame>`, and the other side will hold:

* If it's an interface, it'll also hold `Bidi<Frame, Frame>`
* If it's a local service, it'll hold `Bidi<ToService, ToClient>`

Comment (jamesmunns) [reply source/mgnp-pitch/src/lib.rs:67]: Specifically, it'll look kinda like this under the hood:

```
Bidi<Frame, Frame>    Bidi<TS, TC>
  rx <- (ser)   <- TC <-  tx
  tx -> (deser) -> TS ->  rx
```

The one nit is that at least in my current sketch, a conn drop means that the Bidi (or something holding it) needs to be be able to create a "close" Frame with the right ID.

Comment (hawkw) [reply source/mgnp-pitch/src/lib.rs:67]: hmm, i think we still need to be doing a (tiny bit of) serializing either here or in the conn-table.
this is because, although each conn in the conn-table has a bidi of frames, in
my current implementation, the conn table will be responsible for selecting over
those bidis and determining their connection ID, so we will need to be writing
the ID here?

Comment (hawkw) [reply source/mgnp-pitch/src/lib.rs:67]: that's sort of the tricky thing here that i have yet to figure out: if we're receiving a `&[u8]` frame from the bidi, how do we prefix that with a link ID without having to realloc? i think we may want to change the `Wire` trait so it's not a thing of `Frame`s but a thing you can write bytes to, so you can write the link ID and then the payload.

File: source/mgnp-pitch/Cargo.toml
Comment (hawkw) [inline source/mgnp-pitch/Cargo.toml:1]: we might wanna just rename this crate to `mgnp`, since it now contains a mostly worky proto impl...

File: source/mgnp-pitch/tests/support.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+                let data = data.to_vec().unwrap();
+                postcard::to_allocvec(&InboundMessage::Data {
+                    local_id,
+                    remote_id,
+                    data: &data[..],
+                })
+                .unwrap()
+            }
+        };
+
```
Comment (hawkw) [inline source/mgnp-pitch/tests/support.rs:203]: the fact that we have to serialize into a vec here rather than writing directly to the "wire" (which might be e.g. a byte pipe) is a huge deficiency of the current impl. we need to fix this before we're able to use this stuff in "real life" --- the outbound message should be able to be serialized to a byte slice or directly to an `io::Write`/`embedded_io::Write`/etc...

Comment (jamesmunns) [reply source/mgnp-pitch/tests/support.rs:203]: I *might* have a hack/an idea for this, in one of the branches where I was trying to get no-alloc stuff working. Let me look at it with fresh eyes tomorrow.
